### PR TITLE
Alerting: Respect "For" Duration when NoData set to Alerting

### LINKED ISSUE #56067: Respect "For" Duration for NoData alerts
### Discussed in https://github.com/grafana/grafana/discussions/48849

<div type='discussions-op-text'>

<sup>Originally posted by **joeblubaugh** May  9, 2022</sup>
Grafana 8 alerting lets you define states for query execution errors and no data conditions. You can have them alert, have them be "OK" or have them notify you with the states "Error" and "No Data".

In the NoData condition mapping code there's a switch:
```go
func (a *State) resultNoData(alertRule *ngModels.AlertRule, result eval.Result) {
	a.Error = result.Error

	if a.StartsAt.IsZero() {
		a.StartsAt = result.EvaluatedAt
	}
	a.setEndsAt(alertRule, result)

	switch alertRule.NoDataState {
	case ngModels.Alerting:
		a.State = eval.Alerting
	case ngModels.NoData:
		a.State = eval.NoData
	case ngModels.OK:
		a.State = eval.Normal
	}
}
```
This switch will immediately set the value of an alert to Alerting on the first NoData evaluation. I guessed that it would wait for the "For" period in the same way that Error alerting does. Should it do that?</div>

Answered by @yuri-tceretian :
<div type='discussions-op-text'>

The same way we https://github.com/grafana/grafana/pull/47052 for option Alerting for Error state, this option should consider the rule's for parameter.

[grafana/pkg/services/ngalert/state/state.go](https://github.com/grafana/grafana/blob/4b417c8f3e7d77b6d790d77432bb8e60f634fdef/pkg/services/ngalert/state/state.go#L128-L138)

Lines 128 to 138 in [4b417c8](https://github.com/grafana/grafana/commit/4b417c8f3e7d77b6d790d77432bb8e60f634fdef)
 default: 
```go
 	// For is observed when Alerting is chosen for the alert state 
 	// if execution error or timeout. 
 	if execErrState == eval.Alerting && alertRule.For > 0 { 
 		a.State = eval.Pending 
 	} else { 
 		a.State = execErrState 
 	} 
 	a.StartsAt = result.EvaluatedAt 
 	a.setEndsAt(alertRule, result) 
 }
```
</div>

### PR DESCRIPTION
**Pull request by JacobsonMT**
**What is this feature?**

This change modifies `resultNoData` to be more inline with the logic of the other state handlers.

The main effects of this are:

1) NoData states with NoDataState config set to Alerting will respect "For" duration.
2) Prevents zero value in StartsAt and EndsAt for alerts that have only even been in normal state. This includes state transitions from NoDataState=OK and ExecErrState=OK.
3) Better state transition logging.

**Why do we need this feature?**

For users who have `NoDataState=Alerting`, temporarily data availability issues would cause an immediate firing alert. This is different than how this is handled for `ExecErrState=Alerting` which would respect the configured "For" duration.

**Which issue(s) does this PR fix?**:

Fixes #56067
Related #55320

**Special notes for your reviewer:**

image
image

# Release notice breaking change

Grafana Alerting rules with `NoDataState` configuration set to `Alerting` will now respect "For" duration.

### CODE REVIEW

File: pkg/services/ngalert/state/state.go [hunk 1]
```diff
@@ -187,6 +187,10 @@ func NewEvaluationValues(m map[string]eval.NumberValueCapture) map[string]*float
 func resultNormal(state *State, _ *models.AlertRule, result eval.Result, logger log.Logger) {
 	if state.State == eval.Normal {
 		logger.Debug("Keeping state", "state", state.State)
+		if state.StartsAt.IsZero() {
```
Comment (yuri-tceretian) [inline pkg/services/ngalert/state/state.go:190]: I wonder if it's better to set this in cache.getOrCreate https://github.com/grafana/grafana/blob/0c8876c3a2dcf3bf88a83f11be3e676a2e3b133f/pkg/services/ngalert/state/cache.go#L140-L142

**Comment by yuri-tceretian** [comment]
LGTM but I think we can move the code I mentioned above to getOrCreate. Also, I would like @grobinson-grafana to review it too.

**Review by gotjosh — APPROVED** [review]
LGTM

**Review by yuri-tceretian — APPROVED** [review]
LGTM! 👍🏻

**Pull request merged**